File: Week2/Subset/Subset/Deque.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Subset
{
    public class Deque<T> : IEnumerable<T> where T : class
    {
        private QueueItem _firstItem;
        private QueueItem _lastItem;

        private int _size;

        public Deque()
        {
            
        }

        public bool IsEmpty()
        {
            return _size == 0;
        }

        public int Size()
        {
            return _size;
        }

        public void AddFirst(T item)
        {
            CheckIfItemIsNull(item);

            var tmpItem = _firstItem;

            _firstItem = new QueueItem
                             {
                                 Value = item,
                                 Next = tmpItem
                             };

            if (IsEmpty())
                _lastItem = _firstItem;

            _size++;
        }

        private static void CheckIfItemIsNull(T item)
        {
            if (item == null)
                throw new NullReferenceException();
        }

        public void AddLast(T item)
        {
            CheckIfItemIsNull(item);

            var tmpItem = _lastItem;

            _lastItem = new QueueItem
                            {
                                Value = item,                                
                            };

            
            
            if (!IsEmpty())
            {
                tmpItem.Next = _lastItem;
            }
            else
                _firstItem = _lastItem;

            _size++;
        }

        public T RemoveFirst()
        {
            if (IsEmpty())
                throw new NoSuchElementException();

            var tmpItem = _firstItem;

            _firstItem = _firstItem.Next;

            _size--;

            return tmpItem.Value;
        }

        public T RemoveLast()
        {
            if (IsEmpty())
                throw new NoSuchElementException();

            var tmpItem = _lastItem;

            _lastItem = _lastItem.Prev;

            _size--;

            return tmpItem.Value;
        }

        private class QueueItem
        {
            public T Value { get; set; }            
            public QueueItem Next { get; set; }            
        }

        #region IEnumerable

        public IEnumerator<T> GetEnumerator()
        {
            QueueItem iterator = _firstItem;
            while (iterator != null)
            {
                yield return iterator.Value;
                iterator = iterator.Next;
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        #endregion
    }
}

File: Week2/Queues/src/Deque.java
import java.util.Iterator;
import java.util.NoSuchElementException;


public class Deque<Item> implements Iterable<Item> {

    public Deque() {
        firstItem = null;
        lastItem = null;
        size = 0;
    }
    
    private class QueueItem {
        private Item item;
        private QueueItem next;
        private QueueItem prev;
    }
    
    private QueueItem firstItem;
    private QueueItem lastItem;
    private int size;
    
    private void checkIfItemIsNull(Item item) {
        if (item == null)
            throw new NullPointerException();
    }
    
    private void checkIfQueueIsEmpty() {
        if (isEmpty())
            throw new NoSuchElementException();
    }
    
    public boolean isEmpty() {
        return size() == 0;
    }
    
    public int size() {
        return size;
    }
    
    public void addFirst(Item item) {
        checkIfItemIsNull(item);
        
        QueueItem tmp = firstItem;
        
        firstItem = new QueueItem();
        firstItem.item = item;
        firstItem.next = tmp;
        
        if (isEmpty())
            lastItem = firstItem;
        else
            tmp.prev = firstItem;
        
        size++;
    }
    
    public void addLast(Item item) {
        checkIfItemIsNull(item);
        
        QueueItem tmp = lastItem;
        
        lastItem = new QueueItem();
        lastItem.item = item;
        lastItem.prev = tmp;
        
        if (isEmpty())
            firstItem = lastItem;
        else
            tmp.next = lastItem;
        
        size++;
    }
    
    public Item removeFirst() {
        checkIfQueueIsEmpty();
        
        QueueItem tmp = firstItem;
        firstItem = firstItem.next;        
        size--;
        if (size > 0)
            firstItem.prev = null;
        else
            lastItem = null;
        return tmp.item;
    }
    
    public Item removeLast() {
        checkIfQueueIsEmpty();
        
        QueueItem tmp = lastItem;
        lastItem = lastItem.prev;        
        size--;
        if (size > 0)
            lastItem.next = null;
        else
            firstItem = null;
        return tmp.item;        
    }
    
    private class DequeIterator implements Iterator<Item> {

        QueueItem current;        
        
        public DequeIterator() {
            current = firstItem;
        }
        
        @Override
        public boolean hasNext() {
            return current != null;
        }

        @Override
        public Item next() {
            
            if (!hasNext())
                throw new NoSuchElementException();
            
            QueueItem tmp = current;
            current = current.next;
            
            return tmp.item;
        }

        @Override
        public void remove() {
            throw new UnsupportedOperationException();
        }
        
    }
    
    @Override
    public Iterator<Item> iterator() {
        return new DequeIterator();
    }    
}
